File: go/lidy/parser.go
package lidy

import (
	yaml "gopkg.in/yaml.v3"
)

// Builder is a user-implemented input-validation and creation of user objects
type Builder func(input Result) (interface{}, bool, error)

// Parser gathers the properties necessary to perform lidy parsing
type Parser struct {
	schema     YamlFile
	builderMap map[string]Builder
}

type tParserData struct {
	contentFileName string
	ruleSet         map[string]tRule
	ruleTrace       []string
}

// MustMakeParser makes a parser and panic if any error was found
func MustMakeParser(file File, builderMap map[string]Builder) Parser {
	parser, err := MakeParser(file, builderMap)
	if err != nil {
		panic(err)
	}
	return parser
}

// MakeParser tries to make a parser from the given file
func MakeParser(file File, builderMap map[string]Builder) (Parser, error) {
	parser := Parser{
		schema: YamlFile{
			File: file,
		},
		builderMap: builderMap,
	}
	yamlErr := parser.schema.Unmarshal()
	if yamlErr != nil {
		return parser, yamlErr
	}

	// METAPARSING VALIDATION
	// Validate that the provided schema is valid according to the lidy metaschema
	_, metaParsingError := metaParser.parseData(parser.schema)
	if metaParsingError != nil {
		return parser, metaParsingError
	}

	return parser, nil
}

// Parse -- use the parser to check the given YAML file, and produce a Lidy Result.
func (p Parser) Parse(content File) (Result, error) {
	contentData := YamlFile{
		File: content,
	}
	err := contentData.Unmarshal()
	if err != nil {
		return Result{}, err
	}
	return p.parseData(contentData)
}

func (p Parser) parseData(content YamlFile) (Result, error) {
	parserData := tParserData{
		ruleSet:         map[string]tRule{},
		contentFileName: content.Name,
	}

	parserDocument := p.schema.Yaml.Content[0]

	for k := 0; k < len(parserDocument.Content); k += 2 {
		ruleName := parserDocument.Content[k].Value
		builder := p.builderMap[ruleName]
		parserData.ruleSet[ruleName] = tRule{
			name:           ruleName,
			node:           parserDocument.Content[k+1],
			builder:        builder,
			ruleIsMatching: map[*yaml.Node]bool{},
		}
	}

	contentDocument := content.Yaml.Content[0]

	return applyRule(parserData, "main", contentDocument)
}

File: rust/src/parser.rs
use lidy__yaml::{LineCol, Yaml, YamlData};
use std::collections::HashMap;
use std::rc::Rc;

use crate::error::{AnyBoxedError, SimpleError};
use crate::file::File;
use crate::metaparser::{check_rule_set, make_meta_parser_for};
use crate::result::Data;
use crate::rule::{apply_rule, Rule};
use crate::yamlfile::YamlFile;
use crate::LidyResult;

pub struct Parser<'a, TV> {
    pub content_file_name: Rc<str>,
    // The map of rule name to rule content
    pub rule_set: HashMap<Box<str>, Rule>,
    // The map of builder functions for each rule
    pub builder_callback:
        Box<dyn 'a + FnMut(&str, &LidyResult<TV>) -> Result<Data<TV>, AnyBoxedError>>,
    // The stack of the names of the rules
    pub rule_trace: Vec<Box<str>>,
    // Whether this rule is already being processed for a node. This is used
    // to detect infinite loops
    pub rule_is_matching_node: HashMap<RuleNodePair, ()>,
}

#[derive(Clone, Eq, PartialEq, Hash)]
pub struct RuleNodePair {
    pub rule: Box<str>,
    pub node_line_col: LineCol,
}

impl RuleNodePair {
    pub fn new(rule: Box<str>, node: &Yaml) -> Self {
        Self {
            rule,
            node_line_col: node.line_col,
        }
    }
}

impl<'a, TV> Parser<'a, TV> {
    pub fn make(
        file: &Rc<File>,
        builder_callback: Box<
            dyn 'a + FnMut(&str, &LidyResult<TV>) -> Result<Data<TV>, AnyBoxedError>,
        >,
    ) -> Result<Self, AnyBoxedError> {
        let mut schema_file = YamlFile::new(file.clone());
        schema_file.deserialize()?;

        let rule_set = make_rule_set(&schema_file)?;
        let mut parser = Parser {
            content_file_name: file.name.clone().into(),
            rule_set,
            builder_callback,
            rule_trace: Vec::new(),
            rule_is_matching_node: HashMap::new(),
        };

        // METAPARSING VALIDATION
        // Validate that the provided schema is valid according to the lidy metaschema
        {
            let mut meta_parser = make_meta_parser_for::<TV>(&mut parser)?;
            meta_parser.parse_content_yaml_file(&schema_file)?;
        }
        check_rule_set(&mut parser.rule_set)?;

        Ok(parser)
    }

    pub fn parse(&mut self, file: &Rc<File>) -> Result<LidyResult<TV>, AnyBoxedError> {
        let mut yaml_file = YamlFile::new(file.clone());
        yaml_file.deserialize()?;
        self.parse_content_yaml_file(&yaml_file)
    }

    pub fn parse_content_yaml_file(
        &mut self,
        yaml_file: &YamlFile,
    ) -> Result<LidyResult<TV>, AnyBoxedError> {
        apply_rule(self, "main", &yaml_file.yaml)
    }
}

pub fn make_rule_set(yaml_file: &YamlFile) -> Result<HashMap<Box<str>, Rule>, AnyBoxedError> {
    match &yaml_file.yaml.data {
        YamlData::Mapping(mapping) => {
            let mut rule_set = HashMap::new();
            for (key, value) in mapping {
                if let YamlData::String(rule_name) = &key.data {
                    let rule = Rule {
                        name: rule_name.clone().into(),
                        node: value.clone(),
                        is_used: false,
                    };
                    rule_set.insert(Box::from(rule_name.clone()), rule);
                }
            }
            Ok(rule_set)
        }
        _ => Err(SimpleError::from_message(
            format!(
                "The document should be a YAML map, not {:?}",
                yaml_file.yaml.data
            )
            .into(),
        )
        .into()),
    }
}
